### DISCUSSION TITLE
Wheel only includes submodules at top level, sdist is great

### DISCUSSION THREAD

**Discussion by spensmith** [pypa/setuptools#3185] [Q&A] [answered]
I'm trying to use the basic example from the docs, basically just trying to install from wheel instead of from the sdist, but my wheel doesn't have a `proj` directory that I can import, only `pkg1` and `pkg2`. 

**Starting from this directory:** 
```
├── proj
│   ├── __init__.py
│   ├── additional
│   │   └── __init__.py
│   ├── pkg1
│   │   └── __init__.py
│   └── pkg2
│       └── __init__.py
├── pyproject.toml
└── setup.cfg
```

**I tried using this setup.cfg:**

```
[metadata]
name = p1
version = 0.0.1

[options]
packages =
    find:
package_dir =
    =proj

[options.packages.find]
where = proj
include = pkg*
exclude = additional
```

 **To generate this file structure, which is successfully generated in the source distribution:**
```
├── PKG-INFO
├── README.md
├── proj
│   ├── p1.egg-info
│   │   ├── PKG-INFO
│   │   ├── SOURCES.txt
│   │   ├── dependency_links.txt
│   │   └── top_level.txt
│   ├── pkg1
│   │   └── __init__.py
│   └── pkg2
│       └── __init__.py
├── pyproject.toml
└── setup.cfg
```

**But the wheel has this structure:**

```
├── p1-0.0.1.dist-info
│   ├── METADATA
│   ├── RECORD
│   ├── WHEEL
│   └── top_level.txt
├── pkg1
│   └── __init__.py
└── pkg2
    └── __init__.py
```

**This makes it impossible to use:**

```
import proj
```

As there is no such module, only modules `pkg1` and `pkg2`.

I also made a basic repo for this here: https://github.com/spensmith/setuptools-find.

Steps to recreate: 

 1. Clone the base repository here: https://github.com/spensmith/setuptools-find 
 2. `cd setuptools-find` 
 3. `python -m build` 
 4. `mkdir dist/mywheel` 
 5. `unzip dist/p1-0.0.1-py3-none-any.whl -d dist/mywheel` 
 6. `tree dist/mywheel`

**Comment by jwodder** [comment, marked as answer]
The `package_dir` setting is for declaring a directory (usually named "`src`") that contains your Python code; it is not used for the Python package directory itself.  In your case, simply get rid of the `package_dir` setting along with your `[options.packages.find]` section, which (as written here) only makes sense if you're trying to package the code inside `proj` but not `proj` itself.

**Reply by spensmith** [reply]
Thank you. I see my understanding is different. If I'm understanding, you're asking me to use:

```
[metadata]
name = p1
version = 0.0.1

[options]
packages =
    find:

```

Which results in: 
```
dist/mywheel
├── p1-0.0.1.dist-info
│   ├── METADATA
│   ├── RECORD
│   ├── WHEEL
│   └── top_level.txt
├── pkg1
│   └── __init__.py
├── pkg2
│   └── __init__.py
└── proj
    ├── __init__.py
    ├── additional
    │   └── __init__.py
    ├── pkg1
    │   ├── __init__.py
    └── pkg2
        └── __init__.py
```

To me this doesn't work, as we lose the ability to filter and also have extra folders for additional, pkg1, and pkg2 at the top level.

Am I misunderstanding something?

Edit: pkg1 and pkg2 at the top level were because I unzipped a previous wheel at this level, not a result of following @jwodder's direction. So my only unsolved question at this time was "How do I still exclude"?

**Reply by jwodder** [reply]
You can omit `additional` with `[options.packages.find]exclude = proj.additional`.  Those top-level `pkg1` and `pkg2` folders shouldn't be there; try cleaning your repository of everything in `dist/` and `build/` and then rebuild.

**Reply by spensmith** [reply]
I see. Okay sweet, I must have unzipped into that folder, creating the top level pkg1 and pkg2. This is great. Thank you so much @jwodder. 

Did I miss this way of specifying in the documentation? 

My whole intention was to follow how pandas is structured, with a folder inside the repository containing your project code... Anyways, I'm happy to add to the documentation to help others if that might be helpful. 

I kept reading about the src layout, but I wasn't sure which projects use that, and since I'm in the data space I was looking at pandas/numpy as the models.

----

This one has an [example](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html?highlight=src.#configuring-setup-using-setup-cfg-files) like that, but there's no explanation.

And then the following [example](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html?highlight=src.#using-a-src-layout) shows us to use the src with specifying the `package_dir` as `=src`.
